datavault/backend.py

```diff
@@ -216,7 +439,379 @@ def hasMore(self, pos):
+        return row_data, new_pos
+
+    def getDataTranspose(self, limit, start):
+        struct_data, new_pos = self._getData(limit, start)
+        columns = []
+        for idx in range(len(struct_data.dtype)):
+            col = struct_data['f{}'.format(idx)]
+            # Strings are stored as hdf5 vlen objects.  Numpy can't do
+            # variable length strings, so they get encoded as object
+            # arrays by hdf5.  we don't know how to flatten object
+            # arrays so we special case vlen types here and convert
+            # them to lists.  Also, h5py has a bug where when you
+            # index a dataset with a compound type, it loses the
+            # special dtype information, so we pull it directly from
+            # self.dataset.dtype rather than the data returned by
+            # _getData
+            if self.dataset.dtype[idx] == np.object:
+                base_type = h5py.check_dtype(vlen=self.dataset.dtype[idx])
+                if not base_type or not issubclass(base_type, str):
+                    raise RuntimeError("Found object type array, but not vlen str.  Not supported.  This shouldn't happen")
+                col = [base_type(x) for x in col]
+            columns.append(col)
+        columns = tuple(columns)
+        return columns, new_pos
+
+    def _getData(self, limit, start):
+        if limit is None:
+            struct_data = self.dataset[start:]
+        else:
+            struct_data = self.dataset[start:start+limit]
+        return struct_data, start + struct_data.shape[0]
+
+    def __len__(self):
+        return self.dataset.shape[0]
+
+    def hasMore(self, pos):
+        return pos < len(self)
+
+class SimpleHDF5Data(HDF5MetaData):
+    """Basic dataset backed by HDF5 file.
+
+    This is a very simple implementation that only supports a single 2-D dataset
+    of all floats.  HDF5 files support multiple types, multiple dimensions, and
+    a filesystem-like tree of datasets within one file.  Here, the single dataset
+    is stored in /DataVault within the HDF5 file.
+    """
+    def __init__(self, fh):
+        self._file = fh
+        if 'Version' not in self.file.attrs:
+            self.file.attrs['Version'] = np.asarray([2, 0, 0], dtype=np.int32)
+        self.version = np.asarray(self.file.attrs['Version'], dtype=np.int32)
+
+    def initialize_info(self, title, indep, dep):
+        ncol = len(indep) + len(dep)
+        dtype = [('f{}'.format(idx), np.float64) for idx in range(ncol)]
+        if 'DataVault' not in self.file:
+            self.file.create_dataset('DataVault', (0,), dtype=dtype, maxshape=(None,))
+        HDF5MetaData.initialize_info(self, title, indep, dep)
+
+    @property
+    def file(self):
+        return self._file()
+
+    @property
+    def dataset(self):
+        return self.file["DataVault"]
+
+    def addData(self, data):
+        """Adds one or more rows or data from a 2D array of floats."""
+        new_rows = data.shape[0]
```

Comment:
all the addData functions should assume that they'll be passed a numpy array. That wasn't true in the past; we originally wanted this server to work with or without numpy, but at some point decided it's not onerous to require numpy and makes things much faster, but we likely did not update the code to reflect the use of numpy arrays everywhere.